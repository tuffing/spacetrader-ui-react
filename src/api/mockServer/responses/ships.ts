import { Ship, Ships, Manufacturer } from '../../ships';

export const availableShipsSuccessBody: Ships = {
	ships: [
		{
			type: 'EM-MK-III',
			class: 'MK-III',
			maxCargo: 100,
			speed: 3,
			manufacturer: Manufacturer.Electrum,
			plating: 10,
			weapons: 15,
			purchaseLocations: [
				{
					location: 'OE-A1',
					price: 40000
				}
			]
		},
		{
			type: 'EM-MK-II',
			class: 'MK-II',
			maxCargo: 100,
			speed: 2,
			manufacturer: Manufacturer.Electrum,
			plating: 10,
			weapons: 10,
			purchaseLocations: [
				{
					location: 'OE-A1',
					price: 16000
				}
			]
		},
		{
			type: 'ZA-MK-II',
			class: 'MK-II',
			maxCargo: 300,
			speed: 2,
			manufacturer: Manufacturer.Zetra,
			plating: 5,
			weapons: 5,
			purchaseLocations: [
				{
					location: 'OE-A1',
					price: 12800
				},
				{
					location: 'OE-A1-M1',
					price: 12800
				}
			]
		},
		{
			type: 'GR-MK-I',
			class: 'MK-I',
			maxCargo: 300,
			speed: 1,
			manufacturer: Manufacturer.Gravager,
			plating: 10,
			weapons: 5,
			purchaseLocations: [
				{
					location: 'OE-A1-M1',
					price: 2900
				},
				{
					location: 'OE-D2',
					price: 2900
				}
			]
		},
		{
			type: 'ZA-MK-III',
			class: 'MK-III',
			maxCargo: 300,
			speed: 2,
			manufacturer: Manufacturer.Zetra,
			plating: 10,
			weapons: 10,
			purchaseLocations: [
				{
					location: 'OE-A1-M1',
					price: 33000
				}
			]
		},
		{
			type: 'JW-MK-II',
			class: 'MK-II',
			maxCargo: 100,
			speed: 2,
			manufacturer: Manufacturer.Jackshaw,
			plating: 10,
			weapons: 10,
			purchaseLocations: [
				{
					location: 'OE-G4',
					price: 8000
				}
			]
		},
		{
			type: 'JW-MK-I',
			class: 'MK-I',
			maxCargo: 100,
			speed: 2,
			manufacturer: Manufacturer.Jackshaw,
			plating: 5,
			weapons: 5,
			purchaseLocations: [
				{
					location: 'OE-G4',
					price: 1700
				},
				{
					location: 'OE-D2',
					price: 1700
				}
			]
		},
		{
			type: 'GR-MK-III',
			class: 'MK-III',
			maxCargo: 5000,
			speed: 1,
			manufacturer: Manufacturer.Gravager,
			plating: 10,
			weapons: 10,
			purchaseLocations: [
				{
					location: 'OE-D2',
					price: 24000
				}
			]
		},
		{
			type: 'GR-MK-II',
			class: 'MK-II',
			maxCargo: 1000,
			speed: 1,
			manufacturer: Manufacturer.Gravager,
			plating: 10,
			weapons: 5,
			purchaseLocations: [
				{
					location: 'OE-D2',
					price: 10100
				}
			]
		}
	]
};
