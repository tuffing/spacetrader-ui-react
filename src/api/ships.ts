import { performRequest } from './api';

export enum Manufacturer {
	Electrum = 'Electrum',
	Gravager = 'Gravager',
	Jackshaw = 'Jackshaw',
	Zetra = 'Zetra'
}

export interface Ship {
	type: string;
	class: string;
	maxCargo: number;
	speed: number;
	manufacturer: Manufacturer;
	plating: number;
	weapons: number;
	purchaseLocations: { location: string; price: number }[];
}

export interface Ships {
	ships: Ship[];
}

export const availableShips = async (token: string) => {
	const response = await performRequest('GET', `game/ships/`, { Authorization: `Bearer ${token}` });

	return response;
};
