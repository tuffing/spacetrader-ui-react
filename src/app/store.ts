import { configureStore, combineReducers, ThunkAction, Action, getDefaultMiddleware } from '@reduxjs/toolkit';
import userReducer from '../features/user/userSlice';
import shipsReducer from '../features/ships/shipsSlice';
import { persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

const reducers = combineReducers({
	user: userReducer,
	ships: shipsReducer
});

const persistConfig = {
	key: 'root',
	version: 1,
	storage
};

const persistedReducer = persistReducer(persistConfig, reducers);

export const store = configureStore({
	reducer: persistedReducer,
	middleware: getDefaultMiddleware({
		serializableCheck: {
			ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER]
		}
	})
});

export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, Action<string>>;
