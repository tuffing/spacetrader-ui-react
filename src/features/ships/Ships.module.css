.label {
	font-weight: bold;
	margin: 0 1em;
}
