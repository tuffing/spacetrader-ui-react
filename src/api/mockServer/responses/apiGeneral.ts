import { url, AuthTokenBodyFail } from '../../api';

export enum TokenPresets {
	invalidToken = '1234-1234',
	emptyToken = ''
}

export const routeError404: AuthTokenBodyFail = {
	error: {
		message:
			'This route does not exist. Did you type the route correctly? Most route resources are plural except for "/game/", "/game/status" and "/users/:username/token"',
		code: 404
	}
};

export const invalidTokenError401: AuthTokenBodyFail = {
	error: {
		message:
			'Token was invalid or missing from the request. Did you confirm sending the token as a query parameter or authorization header?',
		code: 40101
	}
};
