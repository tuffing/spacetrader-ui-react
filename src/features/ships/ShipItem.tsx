import React from 'react';
import { Ship } from '../../api/ships';
import globalStyles from '../../Styles.module.css';
import shipStyles from './Ships.module.css';

type Props = {
	ship: Ship;
};

export default function ShipItem({ ship }: Props) {
	return (
		<div data-testid="ShipItem" className={globalStyles.row}>
			<span className={shipStyles.label}>Type:</span> {ship.type}
			<span className={shipStyles.label}>Class:</span> {ship.class}
			<span className={shipStyles.label}>Manufacturer:</span> {ship.manufacturer}
			<span className={shipStyles.label}>Max Cargo:</span> {ship.maxCargo}
		</div>
	);
}
