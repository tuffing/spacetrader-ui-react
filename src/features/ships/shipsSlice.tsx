import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { Status } from '../../api/api';
import { availableShips, Ship } from '../../api/ships';

export interface ShipsState {
	availableShips: Ship[];
	status: Status;
	error?: string;
	//eventually put users bought ships here? Or store on User object?
}

export const initialState: ShipsState = {
	availableShips: [],
	status: Status.idle
};

// fetch available ships. @todo change thunk to instead pull this directly from reducers
export const getAvailableShips = createAsyncThunk('ships/getAvailableShips', async (token: string) => {
	return availableShips(token);
});

export const shipsSlice = createSlice({
	name: 'ships',
	initialState,
	reducers: {
		// Use the PayloadAction type to declare the contents of `action.payload`
		setAvailableShips: (state, action: PayloadAction<Ship[]>) => {
			state.availableShips = action.payload;
		},
		setStatus: (state, action: PayloadAction<Status>) => {
			state.status = action.payload;
		}
	},
	extraReducers: (builder) => {
		builder
			.addCase(getAvailableShips.pending, (state, action) => {
				state.status = Status.loading;
			})
			.addCase(getAvailableShips.fulfilled, (state, action) => {
				state.availableShips = action.payload.ships;

				state.status = Status.idle;
			})
			.addCase(getAvailableShips.rejected, (state, action) => {
				state.status = Status.failed;
				state.error = action.error.message;
			});
	}
});

export const { setAvailableShips, setStatus } = shipsSlice.actions;

export default shipsSlice.reducer;
