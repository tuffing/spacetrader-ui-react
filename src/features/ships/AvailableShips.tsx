import React from 'react';
import { RootState } from '../../app/store';
import { useSelector } from 'react-redux';
import ShipItem from './ShipItem';
import globalStyles from '../../Styles.module.css';

export default function AvailableShips() {
	const availableShips = useSelector((state: RootState) => state.ships.availableShips);

	const renderedListItems = availableShips.map((ship) => {
		return <ShipItem key={ship.type} ship={ship} />;
	});

	return (
		<div className={globalStyles.container}>
			<div className={globalStyles.header}>Available Ships</div>

			<div aria-label="Ships available to buy">{renderedListItems}</div>
		</div>
	);
}
