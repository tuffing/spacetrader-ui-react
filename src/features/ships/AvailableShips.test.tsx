import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import { store } from '../../app/store';
import AvailableShips from './AvailableShips';
import { setAvailableShips, setStatus } from './shipsSlice';
import { availableShipsSuccessBody } from '../../api/mockServer/responses/ships';

describe('Available Ships Component', () => {
	beforeAll(() => {
		store.dispatch(setAvailableShips(availableShipsSuccessBody.ships));
	});

	it('Available Ships Snapshot', () => {
		const availableShips = render(
			<Provider store={store}>
				<AvailableShips />
			</Provider>
		);

		expect(availableShips).toMatchSnapshot();
	});

	it('All Availabe ships rendered', () => {
		const availableShips = render(
			<Provider store={store}>
				<AvailableShips />
			</Provider>
		);

		expect(screen.getAllByTestId('ShipItem').length.toString()).toMatch(
			availableShipsSuccessBody.ships.length.toString()
		);

		store.dispatch(setAvailableShips([]));

		expect(screen.queryAllByTestId('ShipItem').length.toString()).toMatch('0');
	});

	// @todo connection error, filters?
});
