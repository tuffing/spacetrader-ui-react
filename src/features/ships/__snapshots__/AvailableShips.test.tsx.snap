// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Available Ships Component Available Ships Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="header"
        >
          Available Ships
        </div>
        <div
          aria-label="Ships available to buy"
        >
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            EM-MK-III
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-III
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Electrum
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            100
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            EM-MK-II
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-II
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Electrum
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            100
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            ZA-MK-II
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-II
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Zetra
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            300
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            GR-MK-I
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-I
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Gravager
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            300
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            ZA-MK-III
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-III
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Zetra
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            300
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            JW-MK-II
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-II
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Jackshaw
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            100
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            JW-MK-I
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-I
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Jackshaw
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            100
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            GR-MK-III
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-III
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Gravager
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            5000
          </div>
          <div
            class="row"
            data-testid="ShipItem"
          >
            <span
              class="label"
            >
              Type:
            </span>
             
            GR-MK-II
            <span
              class="label"
            >
              Class:
            </span>
             
            MK-II
            <span
              class="label"
            >
              Manufacturer:
            </span>
             
            Gravager
            <span
              class="label"
            >
              Max Cargo:
            </span>
             
            1000
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="header"
      >
        Available Ships
      </div>
      <div
        aria-label="Ships available to buy"
      >
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          EM-MK-III
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-III
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Electrum
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          100
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          EM-MK-II
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-II
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Electrum
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          100
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          ZA-MK-II
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-II
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Zetra
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          300
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          GR-MK-I
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-I
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Gravager
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          300
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          ZA-MK-III
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-III
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Zetra
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          300
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          JW-MK-II
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-II
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Jackshaw
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          100
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          JW-MK-I
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-I
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Jackshaw
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          100
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          GR-MK-III
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-III
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Gravager
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          5000
        </div>
        <div
          class="row"
          data-testid="ShipItem"
        >
          <span
            class="label"
          >
            Type:
          </span>
           
          GR-MK-II
          <span
            class="label"
          >
            Class:
          </span>
           
          MK-II
          <span
            class="label"
          >
            Manufacturer:
          </span>
           
          Gravager
          <span
            class="label"
          >
            Max Cargo:
          </span>
           
          1000
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
